tests/integration/policy/envoy_ratelimit_test.go

```diff
@@ -0,0 +1,196 @@
+	"istio.io/istio/pkg/test/framework/components/bookinfo"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/istio/ingress"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/framework/label"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/scopes"
+)
+
+var (
+	ist         istio.Instance
+	bookinfoNs  namespace.Instance
+	ratelimitNs namespace.Instance
+	ing         ingress.Instance
+)
+
+func TestRateLimiting_DefaultLessThanOverride(t *testing.T) {
+	framework.
+		NewTest(t).
+		Run(func(ctx framework.TestContext) {
+			res := sendTraffic(ing, t, "Sending traffic...", 300)
+			got429s := float64(res.RetCodes[http.StatusTooManyRequests])
+
+			if got429s <= 0 {
+				t.Errorf("Bad metric value for rate-limited requests (429s): got %f, want at least 1", got429s)
+			}
+		})
+}
+
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite(m).
+		RequireSingleCluster().
+		Label(label.CustomSetup).
+		Setup(istio.Setup(&ist, nil)).
+		Setup(testSetup).
+		Run()
+}
+
+func testSetup(ctx resource.Context) (err error) {
+	bookinfoNs, err = namespace.New(ctx, namespace.Config{
+		Prefix: "istio-bookinfo",
+		Inject: true,
+	})
+	if err != nil {
+		return
+	}
+	if _, err = bookinfo.Deploy(ctx, bookinfo.Config{Namespace: bookinfoNs, Cfg: bookinfo.BookInfo}); err != nil {
+		return
+	}
+
+	ing = ist.IngressFor(ctx.Clusters().Default())
+
+	bookinfoGatewayFile, err := bookinfo.NetworkingBookinfoGateway.LoadGatewayFileWithNamespace(bookinfoNs.Name())
+	if err != nil {
+		return
+	}
+	destinationRule, err := bookinfo.GetDestinationRuleConfigFile(ctx)
+	if err != nil {
+		return
+	}
+	destinationRuleFile, err := destinationRule.LoadWithNamespace(bookinfoNs.Name())
+	if err != nil {
+		return
+	}
+	virtualServiceFile, err := bookinfo.NetworkingVirtualServiceAllV1.LoadWithNamespace(bookinfoNs.Name())
+	if err != nil {
+		return
+	}
+	err = ctx.Config().ApplyYAML(bookinfoNs.Name(),
+		bookinfoGatewayFile,
+		destinationRuleFile,
+		virtualServiceFile)
+	if err != nil {
+		return
+	}
+
+	ratelimitNs, err = namespace.New(ctx, namespace.Config{
+		Prefix: "istio-ratelimit",
+	})
+	if err != nil {
+		return
+	}
+
+	err = setupEnvoyFilter(ratelimitNs, ctx)
+	if err != nil {
+		return
+	}
+
+	yamlContent, err := ioutil.ReadFile("testdata/ratelimitservice.yaml")
+	if err != nil {
+		return
+	}
+
+	err = ctx.Config().ApplyYAML(ratelimitNs.Name(),
+		string(yamlContent),
+	)
+	if err != nil {
+		return
+	}
+	// Wait for redis and ratelimit service to be up.
+	fetchFn := kube.NewPodFetch(ctx.Clusters().Default(), ratelimitNs.Name(), "app=redis")
+	if _, err = kube.WaitUntilPodsAreReady(fetchFn); err != nil {
+		return
+	}
+	fetchFn = kube.NewPodFetch(ctx.Clusters().Default(), ratelimitNs.Name(), "app=ratelimit")
+	if _, err = kube.WaitUntilPodsAreReady(fetchFn); err != nil {
+		return
+	}
+
+	// For envoy filter changes to sync.
```

Comment:
Do we need to wait for 15s here? I think redis and rate limit pod already take much longer time to become ready.

Reply:
This is because envoyfilter changes didn't sync up in that much time.. get RLSE errors initially..

Reply:
Had to change to 60s for stability when changed to use echo app and thus applying envoy ratelimting config on sidecars...

Reply:
This seems like a bug? We should not need to sleep? I am not sure how it could take 60s to apply

Reply:
it seems to take 15s to 60s... so for stability I have kept it to 60s..

Reply:
Is it slow in envoy or in pilot?

Reply:
Is this time taking for envoy filter configuration to reach proxy? This is not expected. Or is this something else? How did you measure it?

Reply:
It is slow for envoy sidecar to reach ratelimit service... 

do we have metrics on envoy filter getting applied ? I can try to debug when they become available..

Reply:
Tried couple of things like using the cluster added by pilot, changing the dns_refresh_rate for the rate_limit_cluster.. but still need some sleep...